plugins/cakephp3/cakephp3.plugin.zsh

```diff
@@ -1,19 +1,38 @@
 # CakePHP 3 basic command completion
 _cakephp3_get_command_list () {
-	cakephp3commands=($(bin/cake completion commands));printf "%s\n" "${cakephp3commands[@]}"
+	bin/cake Completion commands
+}
+
+_cakephp3_get_sub_command_list () {
+	bin/cake Completion subcommands ${words[2]}
+}
+
+_cakephp3_get_3rd_argument () {
+	bin/cake ${words[2]} ${words[3]} | grep '\-\ '| awk '{print $2}'
 }
 
 _cakephp3 () {
   if [ -f bin/cake ]; then
-    compadd `_cakephp3_get_command_list`
+		if (( CURRENT == 2 )); then
+			compadd `_cakephp3_get_command_list`
+		fi
+		if (( CURRENT == 3 )); then
+			compadd `_cakephp3_get_sub_command_list`
+		fi
+		if (( CURRENT == 4 )); then
+			if [[ ${has3rdargument[(i)${words[3]}]} -le ${#has3rdargument} ]] ; then ;
+			 compadd `_cakephp3_get_3rd_argument`
+			fi
+		fi
   fi
 }
 
+has3rdargument=("all" "controller" "fixture" "model" "template")
```

Comment:
It's cleaner defining this variable inside the `_cakephp3` function. It should work as well.